Implement Jörmungandr TransactionLayer

Anviking opened the pull request:
# Issue Number

#219 

~~:warning: base branch is `KtorZ/transaction-id-golden-tests` :warning:~~

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have implemented a preliminary `TransactionLayer`


# Comments

- [x] Golden tests are wip (not pushed). ~I'd call for merging this as soon as possible still.~

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

Anviking committed:
Implement jörmungandr mkStdTx (first draft)

Anviking commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs:
```diff
@@ -1,19 +1,67 @@
-newTransactionLayer = TransactionLayer
-    { mkStdTx = error "TODO: See http-bridge as starting point"
+newTransactionLayer
+    :: forall n . Hash "Block0Hash"
```
@KtorZ 

> I am not sure about this type ... And, does this really have to be the hash from the first block header?

To my knowledge — yes

Anviking replied:
https://github.com/input-output-hk/rust-cardano/blob/e5d974f7bedeb00c9c9d688ac66094a34bf8f40d/chain-impl-mockchain/src/transaction/witness.rs#L60-L62

paweljakubas commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs:
```diff
@@ -1,19 +1,73 @@
 
 import Prelude
 
+import Cardano.Wallet.Jormungandr.Compatibility
```
let's be more specific here : 
```
import Cardano.Wallet.Jormungandr.Compatibility
    ( Jormungandr )
```

paweljakubas commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs:
```diff
@@ -1,19 +1,73 @@
+-- Rust implementation:
+--
+-- https://github.com/input-output-hk/rust-cardano/blob/13af7aa08cdac8efc9dc8b0f7ff9162d28b2d479/chain-impl-mockchain/src/transaction/witness.rs#L60-L61
+witnessUtxoData :: Hash "Block0Hash" -> Hash "Tx" -> WitnessData
```
minor stuff, but maybe we would like to have all functions below in `where` of newTransactionLayer` ???

Anviking replied:
I was thinking one _might_ want to export and test individually. And `witnessUtxoData` will be (one of several) "constructors" for `WitnessData`, so I think they should be on the same level as the `WitnessData` declaration.

Happy to surrender though. 

cc @KtorZ 

> Why the WitnessData indirection here? Why not have the hashes as arguments directly?

KtorZ committed:
minor style adjustment, mostly removing extra newlines

Anviking committed:
Make 'TxWitness' a mere wrapper around a 'ByteString'

This makes it more flexible to manipulate 'TxWitness'
across different backend target. The application layer
only regards it as a bunch of bytes anyway.

Co-authored-by: KtorZ <matthias.benkort@gmail.com>

akegalj commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:
```diff
@@ -203,21 +212,26 @@ getMessage = label "getMessage" $ do
+putMessage :: Message -> Put
+putMessage m = withSizeHeader16be $ case m of
+    Transaction (tx, wits) -> do
+        putWord8 2
```
we should probably add a comment explaining why constant 2 (or reference to the actuall rust api line containing this info)

akegalj commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:
```diff
@@ -244,37 +258,43 @@ putSignedTransaction (tx, witnesses) = do
     case witness of
-        PublicKeyWitness xPub (Hash sig) -> do
+        PublicKeyWitness _xPub (Hash sig) -> do
             -- Witness sum type:
```
we might want to define sum type that would describe the constants in this module, at least for documentation

akegalj commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:
```diff
@@ -244,37 +258,43 @@ putSignedTransaction (tx, witnesses) = do
+            --   * 3 for multisig
             putWord8 1
-            putByteString xPub
+            --putByteString xPub
```
But then old address scheme won't work, right? (as it expects xPub in its format)

akegalj commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs:
```diff
@@ -244,37 +258,43 @@ putSignedTransaction (tx, witnesses) = do
         -- NOTE: special value 0xff indicates account spending
         -- only old utxo/address scheme supported for now
         putWord8 $ fromIntegral inputIx
+        putWord64be 14 -- TODO: We need input value!
```
I am not sure I understand the idea behind this ?

EDIT: aha - I see https://github.com/input-output-hk/rust-cardano/blob/13af7aa08cdac8efc9dc8b0f7ff9162d28b2d479/chain-impl-mockchain/src/transaction/transfer.rs#L21 :+1:

akegalj commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Compatibility.hs:
```diff
@@ -84,7 +79,7 @@ genesis = BlockHeader
 -- https://github.com/input-output-hk/rust-cardano/blob/e5d974f7bedeb00c9c9d688ac66094a34bf8f40d/chain-impl-mockchain/src/transaction/transaction.rs#L115-L119
 instance TxId (Jormungandr n) where
-    txId = blake2b256 . putTokenTransfer
+    txId = blake2b256 . putTxBody
```
:+1: https://github.com/input-output-hk/rust-cardano/blob/e5d974f7bedeb00c9c9d688ac66094a34bf8f40d/chain-impl-mockchain/src/transaction/transaction.rs#L117

akegalj commented on lib/jormungandr/src/Cardano/Wallet/Jormungandr/Network.hs:
```diff
@@ -185,6 +199,20 @@ mkJormungandrLayer mgr baseUrl = JormungandrLayer
                         (Just count)
                 left ErrGetDescendantsNetworkUnreachable <$> defaultHandler ctx x
+    , postSignedTx = \tx -> ExceptT $ do
+        let e0 = "?"
```
I can't find where this was taken from?

Anviking replied:
Sorry, for the unnecessary cognitive load by including this commit. (I shouldn't have!) I copied it from `http-bridge` (for testing purposes) where `?` was an error message. I have not at all looked into what kind of errors jormungandr returns. Dropping it now.

akegalj commented on lib/jormungandr/test/unit/Cardano/Wallet/Jormungandr/TransactionSpec.hs:
```diff
@@ -171,3 +211,150 @@ instance Arbitrary (Hash "Tx") where
+            let keyFrom _ = Just (key, pwd)
+            let inps = pure
+                    ( txin
+                    , TxOut (Address "used for lookup") (Coin 14)
```
address content is not important?

akegalj reviewed (commented):
It mostly looks good to me. There are few places where I am not sure in the correctness (I wasn't able to find the corresponding place in rust code)

akegalj committed:
Fix build errors

Co-authored-by: KtorZ <matthias.benkort@gmail.com>

KtorZ committed:
replace 'whileM' with 'many' and remove unecessary 'FromBinary' indirection

KtorZ committed:
review goldent tests for comparing tx with jcli

KtorZ committed:
fix witness creation: use txId from each input!

KtorZ committed:
re-generate nix machinery

KtorZ committed:
review Jormungandr BinarySpec module (favor hex encoding + remove external file)

KtorZ committed:
cleanup dead-code even further in Jormungandr Binary

KtorZ committed:
add decode test for Jormungandr genesis block on Mainnet
